File: typescript/tsfind/src/findoptions.ts
/*
 * findoptions.js
 *
 * defines the set of find options and provides functionality to define find settings from them
 */

'use strict';

import * as config from './config';
import {ArgToken} from "./argtoken";
import {ArgTokenizer} from './argtokenizer';
import {ArgTokenType} from "./argtokentype";
import {FileUtil} from './fileutil';
import {FindError} from "./finderror";
import {FindOption} from './findoption';
import {FindSettings} from './findsettings';
import {SortUtil} from "./sortutil";

interface StringActionMap {
    [key: string]: any
}

export class FindOptions {
    options: FindOption[];
    boolActionMap: StringActionMap;
    stringActionMap: StringActionMap;
    intActionMap: StringActionMap;
    argTokenizer: ArgTokenizer;

    constructor() {
        this.options = [];

        this.boolActionMap = {
            'archivesonly':
                (b: boolean, settings: FindSettings) => { settings.archivesOnly = b; },
            'colorize':
                (b: boolean, settings: FindSettings) => { settings.colorize = b; },
            'debug':
                (b: boolean, settings: FindSettings) => { settings.debug = b; },
            'excludearchives':
                (b: boolean, settings: FindSettings) => { settings.includeArchives = !b; },
            'excludehidden':
                (b: boolean, settings: FindSettings) => { settings.includeHidden = !b; },
            'followsymlinks':
                (b: boolean, settings: FindSettings) => { settings.followSymlinks = b; },
            'includearchives':
                (b: boolean, settings: FindSettings) => { settings.includeArchives = b; },
            'includehidden':
                (b: boolean, settings: FindSettings) => { settings.includeHidden = b; },
            'help':
                (b: boolean, settings: FindSettings) => { settings.printUsage = b; },
            'nocolorize':
                (b: boolean, settings: FindSettings) => { settings.colorize = !b; },
            'nofollowsymlinks':
                (b: boolean, settings: FindSettings) => { settings.followSymlinks = !b; },
            'noprintdirs':
                (b: boolean, settings: FindSettings) => { settings.printDirs = !b; },
            'noprintfiles':
                (b: boolean, settings: FindSettings) => { settings.printFiles = !b; },
            'norecursive':
                (b: boolean, settings: FindSettings) => { settings.recursive = !b; },
            'printdirs':
                (b: boolean, settings: FindSettings) => { settings.printDirs = b; },
            'printfiles':
                (b: boolean, settings: FindSettings) => { settings.printFiles = b; },
            'recursive':
                (b: boolean, settings: FindSettings) => { settings.recursive = b; },
            'sort-ascending':
                (b: boolean, settings: FindSettings) => { settings.sortDescending = !b; },
            'sort-caseinsensitive':
                (b: boolean, settings: FindSettings) => { settings.sortCaseInsensitive = b; },
            'sort-casesensitive':
                (b: boolean, settings: FindSettings) => { settings.sortCaseInsensitive = !b; },
            'sort-descending':
                (b: boolean, settings: FindSettings) => { settings.sortDescending = b; },
            'verbose':
                (b: boolean, settings: FindSettings) => { settings.verbose = b; },
            'version':
                (b: boolean, settings: FindSettings) => { settings.printVersion = b; }
        };

        this.stringActionMap = {
            'in-archiveext':
                (s: string, settings: FindSettings) => { settings.addInArchiveExtensions(s); },
            'in-archivefilepattern':
                (s: string, settings: FindSettings) => { settings.addInArchiveFilePatterns(s); },
            'in-dirpattern':
                (s: string, settings: FindSettings) => { settings.addInDirPatterns(s); },
            'in-ext':
                (s: string, settings: FindSettings) => { settings.addInExtensions(s); },
            'in-filepattern':
                (s: string, settings: FindSettings) => { settings.addInFilePatterns(s); },
            'in-filetype':
                (s: string, settings: FindSettings) => { settings.addInFileTypes(s); },
            'maxlastmod':
                (s: string, settings: FindSettings) => { settings.maxLastModFromString(s); },
            'minlastmod':
                (s: string, settings: FindSettings) => { settings.minLastModFromString(s); },
            'out-dirpattern':
                (s: string, settings: FindSettings) => { settings.addOutDirPatterns(s); },
            'out-archiveext':
                (s: string, settings: FindSettings) => { settings.addOutArchiveExtensions(s); },
            'out-archivefilepattern':
                (s: string, settings: FindSettings) => { settings.addOutArchiveFilePatterns(s); },
            'out-ext':
                (s: string, settings: FindSettings) => { settings.addOutExtensions(s); },
            'out-filepattern':
                (s: string, settings: FindSettings) => { settings.addOutFilePatterns(s); },
            'out-filetype':
                (s: string, settings: FindSettings) => { settings.addOutFileTypes(s); },
            'path':
                (s: string, settings: FindSettings) => { settings.paths.push(s); },
            'sort-by':
                (s: string, settings: FindSettings) => { settings.sortBy = SortUtil.nameToSortBy(s); }
        };

        this.intActionMap = {
            'maxdepth':
                (i: number, settings: FindSettings) => { settings.maxDepth = i; },
            'maxsize':
                (i: number, settings: FindSettings) => { settings.maxSize = i; },
            'mindepth':
                (i: number, settings: FindSettings) => { settings.minDepth = i; },
            'minsize':
                (i: number, settings: FindSettings) => { settings.minSize = i; },
        };

        this.setOptionsFromJsonFile();
        this.argTokenizer = new ArgTokenizer(this.options);
    }

    private static optCmp(o1: FindOption, o2: FindOption) {
        const a: string = o1.sortArg;
        const b: string = o2.sortArg;
        return a.localeCompare(b);
    }

    // setOptionsFromJsonFile
    private setOptionsFromJsonFile(): void {
        const json = FileUtil.getFileContentsSync(config.FIND_OPTIONS_JSON_PATH);
        const obj = JSON.parse(json);
        if (Object.prototype.hasOwnProperty.call(obj, 'findoptions') && Array.isArray(obj['findoptions'])) {
            obj['findoptions'].forEach(fo => {
                const longArg = fo['long'];
                let argType = ArgTokenType.Unknown;
                if (this.boolActionMap[longArg]) {
                    argType = ArgTokenType.Bool;
                } else if (this.stringActionMap[longArg]) {
                    argType = ArgTokenType.Str;
                } else if (this.intActionMap[longArg]) {
                    argType = ArgTokenType.Int;
                }
                let shortArg = '';
                if (Object.prototype.hasOwnProperty.call(fo, 'short')) {
                    shortArg = fo['short'];
                }
                const desc = fo['desc'];
                this.options.push(new FindOption(shortArg, longArg, desc, argType));
            });
            // add the path option (not in the json file)
            this.options.push(new FindOption('', 'path', '', ArgTokenType.Str));
        } else throw new Error(`Invalid findoptions file: ${config.FIND_OPTIONS_JSON_PATH}`);
        this.options.sort(FindOptions.optCmp);
    }

    private updateSettingsFromArgTokens(settings: FindSettings, argTokens: ArgToken[]): Error | undefined {
        let err: Error | undefined = undefined;
        for (const argToken of argTokens) {
            if (err) break;
            if (argToken.type === ArgTokenType.Bool) {
                if (typeof argToken.value === 'boolean') {
                    this.boolActionMap[argToken.name](argToken.value, settings);
                } else {
                    err = new FindError(`Invalid value for option: ${argToken}`);
                }
            } else if (argToken.type === ArgTokenType.Str) {
                if (argToken.name === 'settings-file') {
                    err = this.updateSettingsFromFile(settings, argToken.value);
                } else if (typeof argToken.value === 'string') {
                    this.stringActionMap[argToken.name](argToken.value, settings);
                } else if (typeof argToken.value === 'object' && Array.isArray(argToken.value)) {
                    argToken.value.forEach(s => {
                        if (typeof s === 'string') {
                            this.stringActionMap[argToken.name](s, settings);
                        } else {
                            err = new FindError(`Invalid value for option: ${argToken}`);
                        }
                    });
                } else {
                    err = new FindError(`Invalid value for option: ${argToken}`);
                }
            } else if (argToken.type === ArgTokenType.Int) {
                if (typeof argToken.value === 'number') {
                    this.intActionMap[argToken.name](argToken.value, settings);
                } else {
                    err = new FindError(`Invalid value for option: ${argToken}`);
                }
            } else {
                err = new FindError(`Invalid option: ${argToken}`);
            }
        }
        return err;
    }

    public updateSettingsFromJson(settings: FindSettings, json: string): Error | undefined {
        let { err, argTokens } = this.argTokenizer.tokenizeJson(json);
        if (!err) {
            err = this.updateSettingsFromArgTokens(settings, argTokens);
        }
        return err;
    }

    public updateSettingsFromFile(settings: FindSettings, filePath: string): Error | undefined {
        let { err, argTokens } = this.argTokenizer.tokenizeFile(filePath);
        if (!err) {
            err = this.updateSettingsFromArgTokens(settings, argTokens);
        }
        return err;
    }

    public updateSettingsFromArgs(settings: FindSettings, args: string[]): Error | undefined {
        let { err, argTokens } = this.argTokenizer.tokenizeArgs(args);
        if (!err) {
            err = this.updateSettingsFromArgTokens(settings, argTokens);
        }
        return err;
    }

    public settingsFromArgs(args: string[], cb: (err: Error | undefined, settings: FindSettings) => void): void {
        const settings: FindSettings = new FindSettings();
        // default printFiles to true since it's being run from cmd line
        settings.printFiles = true;
        let err = this.updateSettingsFromArgs(settings, args);
        cb(err, settings);
    }

    private getUsageString(): string {
        let usage: string = 'Usage:\n tsfind [options] <path> [<path> ...]' +
            '\n\nOptions:\n';
        const optStrings: string[] = [];
        const optDescs: string[] = [];
        let longest = 0;
        this.options.forEach((opt: FindOption) => {
            if (opt.longArg === 'path')
                return;
            let optString = ' ';
            if (opt.shortArg)
                optString += '-' + opt.shortArg + ',';
            optString += '--' + opt.longArg;
            if (optString.length > longest)
                longest = optString.length;
            optStrings.push(optString);
            optDescs.push(opt.desc);
        });
        for (let i = 0; i < optStrings.length; i++) {
            let os: string = optStrings[i];
            while (os.length < longest)
                os += ' ';
            usage += os + '  ' + optDescs[i] + '\n';
        }
        return usage;
    }

    public usageWithCode(exitCode: number): void {
        console.log(this.getUsageString());
        process.exit(exitCode);
    }

    public usage(): void {
        this.usageWithCode(0);
    }
}

File: objc/objcfind/Sources/objcfind/FindOptions.m
#import "ArgTokenizer.h"
#import "FileUtil.h"
#import "FindConfig.h"
#import "Regex.h"
#import "FindOptions.h"

@interface FindOptions ()
// private properties
@property NSArray<FindOption*> *findOptions;
@property NSDictionary<NSString*,NSString*> *longArgDict;
@property NSDictionary *boolActionDict;
@property NSDictionary *stringActionDict;
@property NSDictionary *integerActionDict;
@property ArgTokenizer *argTokenizer;

@end

@implementation FindOptions

- (instancetype) init {
    self = [super init];
    if (self) {
        self.findOptions = [self findOptionsFromJson];
        self.longArgDict = [self getLongArgDict];
        self.boolActionDict = [self getBoolActionDict];
        self.stringActionDict = [self getStringActionDict];
        self.integerActionDict = [self getIntegerActionDict];
        self.argTokenizer = [self getArgTokenizer];
    }
    return self;
}

- (NSArray<FindOption*>*) findOptionsFromJson {
    NSMutableString *findOptionsJsonPath = [NSMutableString stringWithString:getXfindSharedPath()];
    [findOptionsJsonPath appendString:@"/findoptions.json"];
    
    if (![[NSFileManager defaultManager] fileExistsAtPath:findOptionsJsonPath]) {
        return nil;
    }
    
    NSMutableArray *findOptions = [[NSMutableArray alloc] initWithCapacity:44];

    NSData *data = [NSData dataWithContentsOfFile:findOptionsJsonPath];
    
    if (NSClassFromString(@"NSJSONSerialization")) {
        NSError *error = nil;
        id jsonObject = [NSJSONSerialization
                         JSONObjectWithData:data
                         options:0
                         error:&error];
        
        if (error) { /* JSON was malformed, act appropriately here */ }
        
        if ([jsonObject isKindOfClass:[NSDictionary class]]) {
            NSArray *findOptionObjects = jsonObject[@"findoptions"];
            for (NSDictionary *findOptionDict in findOptionObjects) {
                NSString *lArg = findOptionDict[@"long"];
                NSString *sArg = findOptionDict[@"short"];
                NSString *desc = findOptionDict[@"desc"];
                FindOption *so = [[FindOption alloc] initWithShortArg:sArg withLongArg:lArg withDesc:desc];
                [findOptions addObject:(FindOption*)so];
            }
        }
    }
    NSArray *sortedOptions = [findOptions sortedArrayUsingComparator:^NSComparisonResult(FindOption *so1, FindOption *so2) {
        return [[so1 sortArg] compare:[so2 sortArg]];
    }];

    return sortedOptions;
}

- (NSDictionary<NSString*,NSString*>*) getLongArgDict {
    NSMutableDictionary *longArgDict = [[NSMutableDictionary alloc] initWithCapacity:68];
    for (FindOption *so in self.findOptions) {
        longArgDict[so.longArg] = so.longArg;
        if (so.shortArg) {
            longArgDict[so.shortArg] = so.longArg;
        }
    }
    // Add path here because it isn't included in findoptions.json
    longArgDict[@"path"] = @"path";
    return [NSDictionary dictionaryWithDictionary:longArgDict];
}

typedef void (^BoolActionBlockType)(BOOL, FindSettings*);

- (NSDictionary<NSString*,BoolActionBlockType>*) getBoolActionDict {
    return @{
        @"archivesonly" : [^void (BOOL b, FindSettings *ss) {
            ss.archivesOnly = b;
            if (b) ss.includeArchives = true;
        } copy],
        @"colorize" : [^void (BOOL b, FindSettings *ss) { ss.colorize = b; } copy],
        @"debug" : [^void (BOOL b, FindSettings *ss) {
            ss.debug = b;
            if (b) ss.verbose = true;
        } copy],
        @"excludearchives" : [^void (BOOL b, FindSettings *ss) { ss.includeArchives = !b; } copy],
        @"excludehidden" : [^void (BOOL b, FindSettings *ss) { ss.includeHidden = !b; } copy],
        @"followsymlinks" : [^void (BOOL b, FindSettings *ss) { ss.followSymlinks = b; } copy],
        @"help" : [^void (BOOL b, FindSettings *ss) { ss.printUsage = b; } copy],
        @"includearchives" : [^void (BOOL b, FindSettings *ss) { ss.includeArchives = b; } copy],
        @"includehidden" : [^void (BOOL b, FindSettings *ss) { ss.includeHidden = b; } copy],
        @"nocolorize" : [^void (BOOL b, FindSettings *ss) { ss.colorize = !b; } copy],
        @"nofollowsymlinks" : [^void (BOOL b, FindSettings *ss) { ss.followSymlinks = !b; } copy],
        @"noprintdirs" : [^void (BOOL b, FindSettings *ss) { ss.printDirs = !b; } copy],
        @"noprintfiles" : [^void (BOOL b, FindSettings *ss) { ss.printFiles = !b; } copy],
        @"norecursive" : [^void (BOOL b, FindSettings *ss) { ss.recursive = !b; } copy],
        @"printdirs" : [^void (BOOL b, FindSettings *ss) { ss.printDirs = b; } copy],
        @"printfiles" : [^void (BOOL b, FindSettings *ss) { ss.printFiles = b; } copy],
        @"recursive" : [^void (BOOL b, FindSettings *ss) { ss.recursive = b; } copy],
        @"sort-ascending" : [^void (BOOL b, FindSettings *ss) { ss.sortDescending = !b; } copy],
        @"sort-caseinsensitive" : [^void (BOOL b, FindSettings *ss) { ss.sortCaseInsensitive = b; } copy],
        @"sort-casesensitive" : [^void (BOOL b, FindSettings *ss) { ss.sortCaseInsensitive = !b; } copy],
        @"sort-descending" : [^void (BOOL b, FindSettings *ss) { ss.sortDescending = b; } copy],
        @"verbose" : [^void (BOOL b, FindSettings *ss) { ss.verbose = b; } copy],
        @"version" : [^void (BOOL b, FindSettings *ss) { ss.printVersion = b; } copy]
    };
}

typedef void (^StringActionBlockType)(NSString*, FindSettings*);

- (NSDictionary<NSString*,StringActionBlockType>*) getStringActionDict {
    return @{
        @"in-archiveext" : ^void (NSString* s, FindSettings *ss) {
            [ss addInArchiveExtension:s];
        },
        @"in-archivefilepattern" : ^void (NSString* s, FindSettings *ss) {
            [ss.inArchiveFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-dirpattern" : ^void (NSString* s, FindSettings *ss) {
            [ss.inDirPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-ext" : ^void (NSString* s, FindSettings *ss) {
            [ss addInExtension:s];
        },
        @"in-filepattern" : ^void (NSString* s, FindSettings *ss) {
            [ss.inFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-filetype" : ^void (NSString* s, FindSettings *ss) {
            [ss addInFileType:s];
        },
        @"maxlastmod" : ^void (NSString* s, FindSettings *ss) {
            [ss setMaxLastModFromString:s];
        },
        @"minlastmod" : ^void (NSString* s, FindSettings *ss) {
            [ss setMinLastModFromString:s];
        },
        @"out-archiveext" : ^void (NSString* s, FindSettings *ss) {
            [ss addOutArchiveExtension:s];
        },
        @"out-archivefilepattern" : ^void (NSString* s, FindSettings *ss) {
            [ss.outArchiveFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-dirpattern" : ^void (NSString* s, FindSettings *ss) {
            [ss.outDirPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-ext" : ^void (NSString* s, FindSettings *ss) {
            [ss addOutExtension:s];
        },
        @"out-filepattern" : ^void (NSString* s, FindSettings *ss) {
            [ss.outFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-filetype" : ^void (NSString* s, FindSettings *ss) {
            [ss addOutFileType:s];
        },
        @"path" : ^void (NSString* s, FindSettings *ss) {
            [ss addPath:s];
        },
        @"sort-by" : ^void (NSString* s, FindSettings *ss) {
            [ss setSortByFromName:s];
        }
    };
}

typedef void (^IntegerActionBlockType)(NSInteger, FindSettings*);

- (NSDictionary<NSString*,IntegerActionBlockType>*) getIntegerActionDict {
    return @{
        @"maxdepth" : [^void (NSInteger i, FindSettings *ss) { ss.maxDepth = i; } copy],
        @"maxsize" : [^void (NSInteger i, FindSettings *ss) { ss.maxSize = i; } copy],
        @"mindepth" : [^void (NSInteger i, FindSettings *ss) { ss.minDepth = i; } copy],
        @"minsize" : [^void (NSInteger i, FindSettings *ss) { ss.minSize = i; } copy]
    };
}

- (ArgTokenizer *) getArgTokenizer {
    NSMutableDictionary *boolDict = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *stringDict = [[NSMutableDictionary alloc] init];
    stringDict[@"path"] = @"path";
    NSMutableDictionary *intDict = [[NSMutableDictionary alloc] init];
    
    for (FindOption *fo in self.findOptions) {
        if (self.boolActionDict[fo.longArg]) {
            boolDict[fo.longArg] = fo.longArg;
            if (fo.shortArg) {
                boolDict[fo.shortArg] = fo.longArg;
            }
        } else if (self.stringActionDict[fo.longArg]) {
            stringDict[fo.longArg] = fo.longArg;
            if (fo.shortArg) {
                stringDict[fo.shortArg] = fo.longArg;
            }
        } else if (self.integerActionDict[fo.longArg]) {
            intDict[fo.longArg] = fo.longArg;
            if (fo.shortArg) {
                intDict[fo.shortArg] = fo.longArg;
            }
        }
    }

    return [[ArgTokenizer alloc] initWithBoolDict:boolDict stringDict:stringDict intDict:intDict];
}

- (void) applyArgTokenToSettings:(ArgToken *)argToken settings:(FindSettings *)settings error:(NSError **)error {
    if (argToken.type == ArgTokenTypeBool) {
        if ([argToken.value isKindOfClass:[NSNumber class]]) {
            NSNumber *num = (NSNumber *)argToken.value;
            BOOL b = [num boolValue];
            void(^block)(BOOL, FindSettings*) = self.boolActionDict[argToken.name];
            block(b, settings);
        } else {
            setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
            return;
        }
    } else if (argToken.type == ArgTokenTypeStr) {
        if ([argToken.value isKindOfClass:[NSString class]]) {
            NSString *s = (NSString *)argToken.value;
            void(^block)(NSString *, FindSettings *) = self.stringActionDict[argToken.name];
            block(s, settings);
        } else if ([argToken.value isKindOfClass:[NSArray class]]) {
            NSArray *arr = (NSArray *)argToken.value;
            for (NSObject *o in arr) {
                if ([o isKindOfClass:[NSString class]]) {
                    NSString *s = (NSString *)argToken.value;
                    void(^block)(NSString *, FindSettings *) = self.stringActionDict[argToken.name];
                    block(s, settings);
                } else {
                    setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
                    return;
                }
            }
        } else {
            setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
            return;
        }
    } else if (argToken.type == ArgTokenTypeInt) {
        if ([argToken.value isKindOfClass:[NSNumber class]]) {
            NSNumber *num = (NSNumber *)argToken.value;
            NSInteger i = [num integerValue];
            void(^block)(NSInteger, FindSettings*) = self.integerActionDict[argToken.name];
            block(i, settings);
        } else {
            setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
            return;
        }
    } else {
        setError(error, [@"Invalid option: " stringByAppendingString:argToken.name]);
        return;
    }
}

- (void) updateSettingsFromArgTokens:(FindSettings *)settings argTokens:(NSArray<ArgToken*> *)argTokens error:(NSError **)error {
    for (ArgToken *argToken in argTokens) {
        if ([argToken.name isEqualToString:@"settings-file"]) {
            [self updateSettingsFromFile:settings filePath:[NSString stringWithFormat:@"%@", argToken.value] error:error];
        } else {
            [self applyArgTokenToSettings:argToken settings:settings error:error];
        }
        if (*error) {
            return;
        }
    }
}

- (void) updateSettingsFromDictionary:(FindSettings *)settings dictionary:(NSDictionary *)dictionary error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeDictionary:dictionary error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (void) updateSettingsFromData:(FindSettings *)settings data:(NSData *)data error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeData:data error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (FindSettings *) settingsFromData:(NSData *)data error:(NSError **)error {
    FindSettings *settings = [[FindSettings alloc] init];
    [self updateSettingsFromData:settings data:data error:error];
    return settings;
}

- (void) updateSettingsFromFile:(FindSettings *)settings filePath:(NSString *)filePath error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeFile:filePath error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (FindSettings *) settingsFromFile:(NSString *)filePath error:(NSError **)error {
    FindSettings *settings = [[FindSettings alloc] init];
    [self updateSettingsFromFile:settings filePath:filePath error:error];
    return settings;
}

- (void) updateSettingsFromArgs:(FindSettings *)settings args:(NSArray *)args error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeArgs:args error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (FindSettings *) settingsFromArgs:(NSArray<NSString*> *)args error:(NSError **)error {
    FindSettings *settings = [[FindSettings alloc] init];
    // default printFiles to true since running as cli
    settings.printFiles = true;
    [self updateSettingsFromArgs:settings args:args error:error];
    return settings;
}

- (NSString*) getUsageString {
    NSMutableString *s = [[NSMutableString alloc] initWithString:@"\nUsage:\n"];
    [s appendString:@" objcfind [options] <path> [<path> ...]\n\n"];
    [s appendString:@"Options:\n"];
    NSMutableArray *optStrings = [NSMutableArray array];
    long longest = 0;
    for (FindOption *so in self.findOptions) {
        NSMutableString *optString = [[NSMutableString alloc] init];
        if (so.shortArg) {
            [optString appendFormat:@"-%@,", so.shortArg];
        }
        [optString appendFormat:@"--%@", so.longArg];
        if ([optString length] > longest) {
            longest = [optString length];
        }
        [optStrings addObject:optString];
    }
    // For some reason, length-specified fields don't work in NSString format strings,
    // so forced to use char * and sprintf
    char *metaString = " %%-%lus  %%s\n";
    char templateString[20];
    sprintf(templateString, metaString, longest);
    for (int i=0; i < [self.findOptions count]; i++) {
        NSString *optString = optStrings[i];
        NSString *optDesc = self.findOptions[i].desc;
        long formatLen = [optString length] + [optDesc length] + 5;
        char formatString[formatLen];
        sprintf(formatString, templateString, [optString UTF8String], [optDesc UTF8String]);
        [s appendString:[NSString stringWithUTF8String:formatString]];
    }
    return [NSString stringWithString:s];
}

- (void) usage:(int)code {
    logMsg([self getUsageString]);
    exit(code);
}

@end
